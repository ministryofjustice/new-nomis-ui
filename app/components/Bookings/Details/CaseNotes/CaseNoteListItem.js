import React from 'react'
import PropTypes from 'prop-types'
import Heading from '@govuk-react/heading'
import uuid from 'uuid/v4'
import Link from '@govuk-react/link'
import { FormattedDate, FormattedDay, FormattedTime } from '../../../intl'
import CaseNoteAmendmentBlock from './CaseNoteAmendmentBlock'
import { caseNoteType, userType } from '../../../../types'
import {
  CaseNote,
  CaseNoteAmendmentButton,
  CaseNoteContent,
  CaseNoteCreation,
  CaseNoteOccurrenceDetails,
  CaseNoteText,
} from './CaseNoteListItem.styles'

const CaseNoteListItem = ({ caseNote, user, offenderNo, caseNoteListReferrer, iepInformation }) => {
  const {
    caseNoteId,
    authorName,
    text,
    amendments,
    occurrenceDateTime,
    creationDateTime,
    subTypeDescription,
    typeDescription,
    authorUserId,
  } = caseNote

  const sameCreator = () => {
    if (!caseNote || !authorUserId || !user.userId) return false

    // case notes service can return username for old notes
    return authorUserId === user.userId.toString() || authorUserId === user.username
  }

  const { cellLocation, offenderName } = iepInformation || {}

  const setPrintIepData = () => {
    const iepSlipData = {
      type: subTypeDescription,
      raisedDate: creationDateTime,
      raisedBy: authorName,
      issuedBy: user.name,
      offenderNo,
      offenderName,
      caseNote: text,
      cellLocation,
      amendments,
    }

    localStorage.setItem('iepSlip', JSON.stringify(iepSlipData))
  }

  const showPrintIEPLink = ['IEP Warning', 'IEP Encouragement'].includes(subTypeDescription)

  return (
    <CaseNote data-qa="case-note">
      <CaseNoteOccurrenceDetails>
        <FormattedDay value={occurrenceDateTime} />
        <FormattedDate value={occurrenceDateTime} />
        <FormattedTime value={occurrenceDateTime} />
        {authorName}
      </CaseNoteOccurrenceDetails>

      <CaseNoteContent>
        <Heading level={2} size="SMALL">
          {typeDescription} | {subTypeDescription}
        </Heading>
        {creationDateTime && (
          <CaseNoteCreation>
            Creation date:{' '}
            <span>
              <FormattedDay value={creationDateTime} /> - <FormattedDate value={creationDateTime} /> -{' '}
              <FormattedTime value={creationDateTime} />
            </span>
          </CaseNoteCreation>
        )}
        <CaseNoteText>{text}</CaseNoteText>
        {amendments &&
          amendments.length > 0 &&
          amendments.map(amendment => <CaseNoteAmendmentBlock amendment={amendment} key={uuid()} />)}
        {sameCreator() && (
          <CaseNoteAmendmentButton
            to={{
              pathname: `/offenders/${offenderNo}/case-notes/${caseNoteId}/amend-case-note`,
              state: { from: caseNoteListReferrer },
            }}
          >
            <button data-qa="make-amendment" type="button" className="button-cancel">
              Make amendment
            </button>
          </CaseNoteAmendmentButton>
        )}
        {showPrintIEPLink && (
          <div>
            <Link noVisitedState href="/iep-slip" target="_blank" onClick={setPrintIepData}>
              Print IEP slip
            </Link>
          </div>
        )}
      </CaseNoteContent>
    </CaseNote>
  )
}

CaseNoteListItem.propTypes = {
  caseNote: caseNoteType.isRequired,
  user: userType.isRequired,
  offenderNo: PropTypes.string.isRequired,
  caseNoteListReferrer: PropTypes.string.isRequired,
  iepInformation: PropTypes.shape({ cellLocation: PropTypes.string, offenderName: PropTypes.string }).isRequired,
}

export default CaseNoteListItem
