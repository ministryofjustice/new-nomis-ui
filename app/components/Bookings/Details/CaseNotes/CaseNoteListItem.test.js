import React from 'react'
import renderer from 'react-test-renderer'
import { IntlProvider } from 'react-intl'
import { MemoryRouter } from 'react-router-dom'
import Link from '@govuk-react/link'
import CaseNoteListItem from './CaseNoteListItem'
import { CaseNoteAmendmentButton } from './CaseNoteListItem.styles'

describe('<CaseNoteListItem />', () => {
  const props = {
    caseNote: {
      agencyId: 'BSI',
      typeDescription: 'Achievements',
      subTypeDescription: 'Gym',
      amendments: [],
      occurrenceDateTime: '2019-01-01T19:00:00',
      text: 'Offender enjoys attending the gym.',
      bookingId: 1,
      authorName: 'User, Test',
      subType: 'GYM',
      type: 'ACHIEVEMENTS',
      creationDateTime: '2019-01-01T21:00:00',
      authorUserId: '1234',
      source: 'INST',
      caseNoteId: 1,
    },
    offenderNo: 'A12345',
    caseNoteListReferrer: '/case-notes',
    iepInformation: {
      cellLocation: 'CELL-123',
      offenderName: 'Test Offender',
    },
  }

  beforeEach(() => {
    props.user = { userId: '5678' }
  })

  it('should match the default snapshot', () => {
    const tree = renderer
      .create(
        <IntlProvider locale="en">
          <CaseNoteListItem {...props} />
        </IntlProvider>
      )
      .toJSON()

    expect(tree).toMatchSnapshot()
  })

  it('should render with amendments snapshot', () => {
    props.caseNote.amendments = [
      {
        creationDateTime: '2019-01-29T21:15:00',
        authorName: 'User, Test',
        additionalNoteText: "Offender says he doesn't like the gym.",
      },
    ]
    const tree = renderer
      .create(
        <IntlProvider locale="en">
          <CaseNoteListItem {...props} />
        </IntlProvider>
      )
      .toJSON()

    expect(tree).toMatchSnapshot()
  })

  it('should render the make amendments button', () => {
    props.user.userId = '1234'

    const tree = renderer.create(
      <IntlProvider locale="en">
        <MemoryRouter>
          <CaseNoteListItem {...props} />
        </MemoryRouter>
      </IntlProvider>
    )
    expect(tree.root.findAllByType(CaseNoteAmendmentButton)).toHaveLength(1)
  })

  it('should render the make amendments button for matching staff', () => {
    props.user.userId = '1234'

    const tree = renderer.create(
      <IntlProvider locale="en">
        <MemoryRouter>
          <CaseNoteListItem {...props} />
        </MemoryRouter>
      </IntlProvider>
    )
    expect(tree.root.findAllByType(CaseNoteAmendmentButton)).toHaveLength(1)
  })

  it('should not render the make amendments button if different user', () => {
    props.user.userId = '12345'

    const tree = renderer.create(
      <IntlProvider locale="en">
        <MemoryRouter>
          <CaseNoteListItem {...props} />
        </MemoryRouter>
      </IntlProvider>
    )
    expect(tree.root.findAllByType(CaseNoteAmendmentButton)).toHaveLength(0)
  })

  it('should not render the make amendments button if user not set', () => {
    props.user = {}

    const tree = renderer.create(
      <IntlProvider locale="en">
        <MemoryRouter>
          <CaseNoteListItem {...props} />
        </MemoryRouter>
      </IntlProvider>
    )
    expect(tree.root.findAllByType(CaseNoteAmendmentButton)).toHaveLength(0)
  })

  it('should render with print IEP slip link', () => {
    props.caseNote.subTypeDescription = 'IEP Warning'
    const tree = renderer.create(
      <IntlProvider locale="en">
        <MemoryRouter>
          <CaseNoteListItem {...props} />
        </MemoryRouter>
      </IntlProvider>
    )

    expect(tree.root.findAllByType(Link)).toHaveLength(1)
  })

  it('should not render with print IEP slip link if not iep', () => {
    props.caseNote.subTypeDescription = 'Something else'
    const tree = renderer.create(
      <IntlProvider locale="en">
        <MemoryRouter>
          <CaseNoteListItem {...props} />
        </MemoryRouter>
      </IntlProvider>
    )

    expect(tree.root.findAllByType(Link)).toHaveLength(0)
  })
})
