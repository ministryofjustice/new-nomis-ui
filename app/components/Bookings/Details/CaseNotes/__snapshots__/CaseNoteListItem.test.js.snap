// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<CaseNoteListItem /> should match the default snapshot 1`] = `
<div
  className="css-1nnv6h2"
  data-qa="case-note"
>
  <div
    className="css-v35mku"
  >
    <span>
      Tuesday
    </span>
    <span>
      01/01/2019
    </span>
    <span>
      9:00 PM
    </span>
    User, Test
  </div>
  <div
    className="css-1bw51u2"
  >
    <h2
      className="e1vcssft0 css-l3100a-StyledHeader-StyledHoc e4u1u2g0"
      size="SMALL"
    >
      Achievements
       | 
      Gym
    </h2>
    <div
      className="css-1x1vvpx"
    >
      Occurrence date:
       
      <span>
        <span>
          Tuesday
        </span>
         - 
        <span>
          01/01/2019
        </span>
         -
         
        <span>
          9:00 PM
        </span>
      </span>
    </div>
    <pre
      className="css-6ghpto"
    >
      Offender enjoys attending the gym.
    </pre>
  </div>
</div>
`;

exports[`<CaseNoteListItem /> should render the with amendments snapshot 1`] = `
<div
  className="css-1nnv6h2"
  data-qa="case-note"
>
  <div
    className="css-v35mku"
  >
    <span>
      Tuesday
    </span>
    <span>
      01/01/2019
    </span>
    <span>
      9:00 PM
    </span>
    User, Test
  </div>
  <div
    className="css-1bw51u2"
  >
    <h2
      className="e1vcssft0 css-l3100a-StyledHeader-StyledHoc e4u1u2g0"
      size="SMALL"
    >
      Achievements
       | 
      Gym
    </h2>
    <div
      className="css-1x1vvpx"
    >
      Occurrence date:
       
      <span>
        <span>
          Tuesday
        </span>
         - 
        <span>
          01/01/2019
        </span>
         -
         
        <span>
          9:00 PM
        </span>
      </span>
    </div>
    <pre
      className="css-6ghpto"
    >
      Offender enjoys attending the gym.
    </pre>
    <div
      className="css-1b30ksf"
    >
      <h3
        className="e1vcssft0 css-l3100a-StyledHeader-StyledHoc e4u1u2g0"
        size="SMALL"
      >
        Amendment
      </h3>
      <pre
        className="css-6ghpto"
      >
        Offender says he doesn't like the gym.
      </pre>
      <div
        className="css-g90fw7"
      >
        <span>
          Tuesday
        </span>
         - 
        <span>
          01/29/2019
        </span>
         -
         
        <span>
          9:15 PM
        </span>
      </div>
      User, Test
    </div>
  </div>
</div>
`;

exports[`<CaseNoteListItem /> should render the with make amendments button snapshot 1`] = `
<div
  className="css-1nnv6h2"
  data-qa="case-note"
>
  <div
    className="css-v35mku"
  >
    <span>
      Tuesday
    </span>
    <span>
      01/01/2019
    </span>
    <span>
      9:00 PM
    </span>
    User, Test
  </div>
  <div
    className="css-1bw51u2"
  >
    <h2
      className="e1vcssft0 css-l3100a-StyledHeader-StyledHoc e4u1u2g0"
      size="SMALL"
    >
      Achievements
       | 
      Gym
    </h2>
    <div
      className="css-1x1vvpx"
    >
      Occurrence date:
       
      <span>
        <span>
          Tuesday
        </span>
         - 
        <span>
          01/01/2019
        </span>
         -
         
        <span>
          9:00 PM
        </span>
      </span>
    </div>
    <pre
      className="css-6ghpto"
    >
      Offender enjoys attending the gym.
    </pre>
    <div
      className="css-1b30ksf"
    >
      <h3
        className="e1vcssft0 css-l3100a-StyledHeader-StyledHoc e4u1u2g0"
        size="SMALL"
      >
        Amendment
      </h3>
      <pre
        className="css-6ghpto"
      >
        Offender says he doesn't like the gym.
      </pre>
      <div
        className="css-g90fw7"
      >
        <span>
          Tuesday
        </span>
         - 
        <span>
          01/29/2019
        </span>
         -
         
        <span>
          9:15 PM
        </span>
      </div>
      User, Test
    </div>
    <a
      className="css-i1xo79"
      href="/offenders/A12345/case-notes/1/amend-case-note"
      onClick={[Function]}
    >
      <button
        className="button-cancel"
        type="button"
      >
        Make amendment
      </button>
    </a>
  </div>
</div>
`;
